docs(ses): Clarify motivation, fix links (#810)
endojs/endo: packages/ses/README.md

# SES

SES is *hardened JavaScript*. SES stands for *fearless cooperation*.
This package is a SES [shim](https://en.wikipedia.org/wiki/Shim_(computing))
for JavaScript features [proposed](https://github.com/tc39/proposal-ses) to
ECMA TC-39.
Hardened JavaScript is highly compatible with ordinary JavaScript, so a

> surprising quantity of existing JavaScript libraries can run on hardened

Reviewer: Why is it surprising to this audience? Where did they get other expectations about SES?

# SES

> SES is a slight modification to the JavaScript language that enables programs

Reviewer 2: I don't like the phrasing "slight modification". I makes it sound like it is not very significant. Rather, I suspect your point is that it is highly compatible, to the point that a tremendous amount of old code, not written to run under SES, nevertheless does compatibly. We should say this explicitly instead.

Reviewer 3: I think SES as a language isn't a good framing. Inventing a new language is trivial (just add or subtract something!). Inventing a solution that makes JS programs secure is really hard, and really valuable. Our framing and marketing should focus on it being a valuable solution for a particular hard problem.

Reviewer: from the SEStival yesterday:

- Proposed definitions
    - A variation of [[JavaScript]] that is largely backwards-compatible that gives programs a leg to stand on in defending their own correctness.
        - Not a language, since it's a tool in the language that removes features from the language.
        - How do you explain adding security to a language by removing features?
    - Making JavaScript easily safe for
        - [[multi-tennancy]]
        - mashups
        - plugins
    - "JavaScript without the parts that others can use to hurt you." - [[Erik Marks]]
    - A runtime enforced compute environment that does better than [[Caja]].
        - We are trying to definitely not include a JS parser in the [[trusted computing base (TCB)]]. (We don't want to trust a parser, anyways).

Reviewer: another take:

SES is hardened JavaScript.

SES provides a platform to safely compose mutually distrusting JavaScript components. It supports the Principle of Least Authority, allowing components to cooperate without being excessively vulnerable to each other.

The JavaScript ecosystem consists of millions of packages, but without SES, using them makes you vulnerable to attacks or even just bugs that they may contain. SES lets you grant each component exactly the authority that it needs, limiting your vulnerability to just that authority. For example, consider [the event-stream incident][1]. With SES, if you don’t grant `event-stream` access to the network, then it cannot steal your assets.

[1]: https://www.theregister.com/2018/11/26/npm_repo_bitcoin_stealer/

> to defend their own correctness, mostly by disabling dangerous features.

Reviewer 3: "Defend their own correctness" is correct, but doesn't appeal very much to me. "Correctness" sounds like it's using formal verification or something arcane and complicated. What about "SES disables dangerous features inherent in JavaScript and makes it possible to run untrusted code securely."

JavaScript without SES is highly reconfigurable at runtime, so programs can

> easily interfere with each other by altering shared objects.

Reviewer 3: "highly reconfigurable" sounds attractive, but it's unattractive.

As far as I can tell, "interfere with each other" used in this way is jargon that only people following Mark would know. 

```suggestion
Running JavaScript without SES leaves your program vulnerable to any other code that executes in the same environment, by altering shared objects.
```

> After "lockdown", JS becomes SES, all of the shared intrinsics and their

Reviewer 3: This is only if you agree with my suggestion above that SES is not a language.

```suggestion
After "lockdown", all of the shared intrinsics and their
```

  no `fetch`). Compartments can be selectively endowed with powerful arguments,
  globals, or modules.

SES safely executes third-party JavaScript 'strict' mode programs in
compartments that have no excess authority in their global scope.
SES runs atop an ES6-compliant platform, enabling safe interaction of
mutually-suspicious code, using object-capability -style programming.

See https://github.com/Agoric/Jessie to see how SES fits into the various
flavors of confined JavaScript execution. And visit

> https://ses-demo.agoric.app/demos/ for a demo.

Reviewer: I'm not a fan of in-your-face URLs.

```suggestion
[the SES demos](https://ses-demo.agoric.app/demos/) to try it out.
```
